### DISCUSSION TITLE
Update table in Prisma

### DISCUSSION THREAD

**Discussion by elwin212** [prisma/prisma#20517] [Q&A]
### Question

I want to update a table, e.g., insert a column into an existing table.

from
```js
// 
model User {
  id                String         @id @default(cuid())
  name              String?
  email             String?        @unique
  subscriptions     Subscription[]
  username          String?        @unique
  accounts          Account[]
  sessions          Session[]
}
```
to
```js
// 
enum RoleType {
   user
   admin
}

model User {
  id                String         @id @default(cuid())
  name              String?
  email             String?        @unique
  subscriptions     Subscription[]
  username          String?        @unique
  role              RoleType
  image             String?
  accounts          Account[]
  sessions          Session[]
}
```

I added a new column role which is a enum to User table.
When I use `npx prisma migrate dev --name add_column_user --create-only`
It said 

> We need to reset the "public" schema at "localhost:5434"
Do you want to continue? All data will be lost.

 
But I don't want to lose my data, is there any way to do this?
> 

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: Windows
- Database: PostgreSQL
- Node.js version: 18.15.0

- prisma                  : 4.16.2
@prisma/client          : 4.16.2
Current platform        : windows

**Comment by janpio** [comment]
Is that all the output it is giving you or is there more and you just shared a small piece of it?

**Reply by elwin212** [reply]
Hi janpio,

That's the whole output I got. Do I have to `npx prisma generate` first then `npx prisma migrate dev --name add_column_user --create-only`?

**Reply by ludralph** [reply]
Hi @elwin212 👋 

When you add a new column to an existing table in Prisma schema, Prisma Migrate will create a migration to apply the schema changes to the database. However, since the migration involves altering the table structure, there is a risk of data loss, especially if the column is defined as `NOT NULL`.

The warning you received when running `npx prisma migrate dev --name add_column_user --create-only` indicates that the migration would require resetting the schema, which would result in data loss.

To avoid data loss, you can modify your schema to make the new column `role` nullable
```prisma
model User {
  id                String         @id @default(cuid())
  name              String?
  email             String?        @unique
  subscriptions     Subscription[]
  username          String?        @unique
  role              RoleType?
  image             String?
  accounts          Account[]
  sessions          Session[]
}

enum RoleType {
   user
   admin
}
```

After making this change, generate a migration and apply it:
```
npx prisma migrate dev --name add_column_user
```

This change will allow you to add the `role` column without requiring a schema reset and without data loss. If you make changes to your schema that involve non-nullable fields and need to preserve existing data, you may need to follow the [expand and contract pattern](https://www.prisma.io/dataguide/types/relational/expand-and-contract-pattern). See [example](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/customizing-migrations#example-use-the-expand-and-contract-pattern-to-evolve-the-schema-without-downtime) of the  pattern.

**Reply by elwin212** [reply]
Hi @ludralph,

Thank you for the reply! At this point, I tried to make the new column `role` nullable and ran `npx prisma generate`.
After that I ran `npx prisma migrate dev --name add_column_user --create-only ` but it still give me the same warning.
In the prisma schema, should I leave the whole database or just the table that I want to modify?
For example
Original schema:
```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider     = "postgresql"
  url          = env("DATABASE_URL")
  relationMode = "prisma"
}

enum VoteType {
  UP
  DOWN
}

enum RoleType {
  user
  admin
}

model Account {
  id                String  @id @default(cuid())
  userId            String
  type              String
  provider          String
  providerAccountId String
  refresh_token     String? @db.Text
  access_token      String? @db.Text
  expires_at        Int?
  token_type        String?
  scope             String?
  id_token          String? @db.Text
  session_state     String?

  user User @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@unique([provider, providerAccountId])
}

model Session {
  id           String   @id @default(cuid())
  sessionToken String   @unique
  userId       String
  expires      DateTime
  user         User     @relation(fields: [userId], references: [id], onDelete: Cascade)
}

model User {
  id                String         @id @default(cuid())
  name              String?
  email             String?        @unique
  subscriptions     Subscription[]
  username          String?        @unique
  image             String?
  accounts          Account[]
  sessions          Session[]
}
```
To :
```
enum RoleType {
  user
  admin
}

model User {
  id                String         @id @default(cuid())
  name              String?
  email             String?        @unique
  subscriptions     Subscription[]
  role              RoleType?
  username          String?        @unique
  image             String?
  accounts          Account[]
  sessions          Session[]
}
```
Sorry, I'm quiet new to prisma.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
